### PR TITLE
Added distribution support to LockedQueue

### PR DESCRIPTION
**Pull request by matthewtff**
Now Emitter finds out a hash of tasks for remote execution
and puts this info to LockedQueue. Also DoRemoteExecute now
has assigned distribution value that is used to pop appropriate
task from queue.

### CODE REVIEW

File: src/base/worker_pool.h [hunk 1]
```diff
@@ -25,6 +24,10 @@ class WorkerPool {
     return WaitUntilShutdown(ZERO_DURATION);
   }
 
+  size_t GetWorkersNumber() const {
```
Comment (matthewtff) [inline src/base/worker_pool.h]: Think it's unused now, should be removed.

**Comment by abyss7** [comment]
Use common name for "shard" across the project, instead of "range", "distribution", etc.

**Comment by abyss7** [comment]
Rebase onto new master.

**Comment by abyss7** [comment]
Use preprocessor hash for sharding, instead of hash from input file basename.

**Comment by abyss7** [comment]
Implement some tests:
1. Check that task goes to the right shard (not by random pick from a queue head).
2. Check that next same task goes to the same shard (not by random pick from a queue head).
3. Check that shards get updated after update from coordinator.

File: build/travis.sh [hunk 2]
```diff
@@ -14,7 +14,7 @@ elif [[ "$(uname)" == Darwin ]]; then
     arch="Mac"
 fi
 
-clang_revision=283753-1
+clang_revision=282487-1
```
Comment (abyss7) [inline build/travis.sh]: Why to downgrade the version?

File: src/base/locked_queue.h [hunk 3]
```diff
@@ -56,41 +63,130 @@ class LockedQueue {
...
+    }
+
+    size_t Size() const THREAD_UNSAFE {
+      size_t index_size = 0u;
+      for (const ShardsList& shards_list : index_) {
+        index_size += shards_list.size();
+      }
+      return index_size;
+    }
+    void Put(typename Queue::iterator item, const ui64 hash) {
```
Comment (abyss7) [inline src/base/locked_queue.h]: `hash` → `shard` ?

File: src/daemon/emitter.cc [hunk 4]
```diff
@@ -394,18 +386,7 @@ void Emitter::DoRemoteExecute(const base::WorkerPool& pool,
...
-      std::get<CONNECTION>(*task)->ReportStatus(status);
-      continue;
-    }
-
-    auto outgoing = std::make_unique<proto::Remote>();
-    if (source.str.empty() && !GenerateSource(incoming, &source)) {
-      failed_tasks_->Push(std::move(*task));
-      continue;
-    }
+    CHECK(!source.str.empty());
```
Comment (abyss7) [inline src/daemon/emitter.cc]: It's a very strong condition. What will happen, if we don't have local cache?

File: src/daemon/emitter.cc [hunk 5]
```diff
@@ -653,13 +640,46 @@ bool Emitter::Reload(const proto::Configuration& conf) {
       optional->Wait();
       return optional->GetValue();
     };
-    Worker worker = std::bind(&Emitter::DoRemoteExecute, this, _1, resolver);
+    const ui32 shard =
+        remote.has_distribution() ? remote.distribution() : 0u;
```
Comment (abyss7) [inline src/daemon/emitter.cc]: `distribution` → `shard` ?

File: src/daemon/emitter.h [hunk 6]
```diff
@@ -24,11 +24,12 @@ class Emitter : public CompilationDaemon {
     MESSAGE = 1,
     SOURCE = 2,
     EXTRA_FILES = 3,
+    HASH = 4,
```
Comment (abyss7) [inline src/daemon/emitter.h]: `HASH` → `SHARD` ?

File: build/travis.sh (hunk 2 above)
Comment (matthewtff) [inline build/travis.sh]: My bad, I'll revert this change

File: src/base/locked_queue.h [hunk 7]
```diff
@@ -56,41 +63,130 @@ class LockedQueue {
...
+    enum InfoIndex {
+      ITERATOR = 0u,
+      HASH = 1u,
+    };
+    enum ReverseInfoIndex {
+      RANGE = 0u,
+      REVERSE_ITERATOR = 1u,
+    };
+
+    Index(LockedQueue<T>* const queue) : queue_(queue) {
```
Comment (matthewtff) [inline src/base/locked_queue.h]: That should be ok, yeah.

File: src/base/locked_queue.h (hunk 3 above)
Comment (matthewtff) [inline src/base/locked_queue.h]: Actually, no. Here and in other places we hash is a hash value computed out of sources. Base on hash value and number of shards a shard could be computed. But anyway we need 2 distinct entities - shards and hashes.

File: src/daemon/emitter.cc (hunk 5 above)
Comment (matthewtff) [inline src/daemon/emitter.cc]: That would also mean a breakage in config files. Is it OK to check both for several releases and remove then?

File: src/daemon/emitter.h (hunk 6 above)
Comment (matthewtff) [inline src/daemon/emitter.h]: Same as in other places about hash -> shard:
It's a calculated hash value, and shard should be computed out of this and current number of shards(that may change).

File: src/daemon/emitter.cc (hunk 4 above)
Comment (matthewtff) [inline src/daemon/emitter.cc]: It should be here in any case. To put to `all_tasks_` we need to now hash of sources of the task. All these fields are filled in `PopulateTask`. There should be no way to put a task to queue without knowing it's hash/sources.

**Pull request closed without merging**